File: src/commands/new.rb
require 'fileutils'

module Commands
  def self.new(args, options)
    raise ArgumentError, "No name provided" if args.length < 1
    raise ArgumentError, "Command 'new' takes only one argument" if args.length > 1
    raise ArgumentError, "Directory #{name} already exists" if Dir.exist?(name)
    raise RuntimeError, "Git is not installed or not in the PATH." if `git --version`.empty?

    name = args.first
    verbose = options[:verbose]

    template_url = "https://github.com/Cillers-com/create-cillers-system"
    puts "Creating new system named '#{name}'..."
    execute_command("git clone #{template_url} #{name}", verbose)

    raise Error, "The directory was not created successfully" unless Dir.exists?(name) 

    Dir.chdir(system_name) do
      FileUtils.rm_rf('.git')
      puts "Removed original Git repository." if verbose

      puts "Initializing new Git repository..."
      execute_command("git init", verbose)
      execute_command("git add .", verbose)
      execute_command("git commit -m 'Initial commit'", verbose)
    end

    puts "New Cillers system '#{name}' successfully created."
  end

  private

  def self.execute_command(command, verbose)
    if verbose
      puts "Executing: #{command}"
      system(command)
    else
      system("#{command} > /dev/null 2>&1")
    end
  end
end

File: src/cli/commands/new.clj
(ns cli.commands.new
  (:require [clojure.string :as str]
            [cli.return :as return]
            [cli.process :as proc]
            [cli.git :as git]
            [cli.file :as file]))

(defn check-args [args]
  (let [name (first args)
        err return/illegal-argument-error]
    (cond
      (empty? args) (err "No name provided")
      (> (count args) 1) (err "Command 'new' takes only one argument")
      (file/exists? name) (err (str "Directory " name " already exists"))
      :else (return/success))))

(defn check-dependencies [] 
  (let [err return/missing-dependency-error]
    (cond 
      (not (git/installed?)) (err "Git is not installed or not in the PATH.")
      :else (return/success))))

(defn command-fn [args options]
  (let [name (first args)
        verbose? (:verbose options)
        repo-url "https://github.com/Cillers-com/create-cillers-system"]
    (return/do-until-error
      [[check-args [args]]
       [check-dependencies []]
       [return/success-println [(str "Creating new system named '" name "'...")]]
       [git/clone [repo-url name verbose?]]
       [git/reset [name verbose?]]
       [return/success-println [(str "New Cillers system '" name "' successfully created.")]]])))

